core/src/code.rs

```diff
@@ -287,33 +287,88 @@ fn check_mut_global_exports(module: &Module) -> Result<(), CodeError> {
     Ok(())
 }
 
+fn check_start_section(module: &Module) -> Result<(), CodeError> {
+    if module.start_section().is_some() {
+        log::debug!("Found start section in contract code, which is not allowed");
+        Err(CodeError::StartSectionExists)
+    } else {
+        Ok(())
+    }
+}
+
+/// Configuration for `Code::try_new_mock_`.
+/// By default all checks enabled.
+pub struct TryNewCodeConfig {
+    /// Instrumentation version
+    pub version: u32,
+    /// Stack height limit
+    pub stack_height: Option<u32>,
+    /// Check exports (wasm contains init or handle exports)
+    pub check_exports: bool,
+    /// Check and canonize stack end
+    pub check_and_canonize_stack_end: bool,
+    /// Check mutable global exports
+    pub check_mut_global_exports: bool,
+    /// Check start section (not allowed for smart contracts)
+    pub check_start_section: bool,
+    /// Make wasmparser validation
+    pub make_validation: bool,
+}
+
+impl Default for TryNewCodeConfig {
+    fn default() -> Self {
+        Self {
+            version: 1,
+            stack_height: None,
+            check_exports: true,
+            check_and_canonize_stack_end: true,
+            check_mut_global_exports: true,
+            check_start_section: true,
+            make_validation: true,
+        }
+    }
+}
+
+impl TryNewCodeConfig {
+    /// New default config without exports checks.
+    pub fn new_no_exports_check() -> Self {
+        Self {
+            check_exports: false,
+            ..Default::default()
+        }
+    }
+}
+
 impl Code {
     /// Create the code by checking and instrumenting `original_code`.
-    pub fn try_new<R, GetRulesFn>(
-        raw_code: Vec<u8>,
-        version: u32,
-        mut get_gas_rules: GetRulesFn,
-        stack_height: Option<u32>,
+    fn try_new_internal<R, GetRulesFn>(
+        original_code: Vec<u8>,
+        get_gas_rules: Option<GetRulesFn>,
+        config: TryNewCodeConfig,
     ) -> Result<Self, CodeError>
     where
         R: Rules,
         GetRulesFn: FnMut(&Module) -> R,
     {
-        wasmparser::validate(&raw_code).map_err(|_| CodeError::Decode)?;
+        if config.make_validation {
+            wasmparser::validate(&original_code).map_err(|_| CodeError::Decode)?;
```

Comment:
How do you feel about introducing CodeError::Validation?
To concrete each error case.

Reply:
Dood idea. Unfortunately this would not make upload_program errors more user friendly because we need to refactor pallet errors, which has one error ProgramConstrunctionFailed for all cases

Reply:
Done